algorithm/cgal/Geometry.h

```diff
@@ -0,0 +1,662 @@
+    }
+    return *transforms_[nth];
+  }
+
+  bool has_plane(int nth) { return is_polygons(nth); }
+  Plane& plane(int nth) { return planes_[nth]; }
+
+  bool has_input_mesh(int nth) { return input_meshes_[nth] != nullptr; }
+
+  const Surface_mesh& input_mesh(int nth) { return *input_meshes_[nth]; }
+
+  bool has_mesh(int nth) { return meshes_[nth] != nullptr; }
+
+  Surface_mesh& mesh(int nth) {
+    if (!has_mesh(nth)) {
+      meshes_[nth].reset(new Surface_mesh);
+    }
+    return *meshes_[nth];
+  }
+
+  bool has_epick_mesh(int nth) { return epick_meshes_[nth] != nullptr; }
+
+  Epick_surface_mesh& epick_mesh(int nth) {
+    if (!has_epick_mesh(nth)) {
+      epick_meshes_[nth].reset(new Epick_surface_mesh);
+      copy_face_graph(mesh(nth), *epick_meshes_[nth]);
+    }
+    return *epick_meshes_[nth];
+  }
+
+  bool has_aabb_tree(int nth) { return aabb_trees_[nth] != nullptr; }
+
+  void update_aabb_tree(int nth) {
+    Surface_mesh& m = mesh(nth);
+    aabb_trees_[nth].reset(new AABB_tree(faces(m).first, faces(m).second, m));
+  }
+
+  AABB_tree& aabb_tree(int nth) {
+    if (!has_aabb_tree(nth)) {
+      update_aabb_tree(nth);
+    }
+    return *aabb_trees_[nth];
+  }
+
+  bool has_on_side(int nth) { return on_sides_[nth] != nullptr; }
+
+  void update_on_side(int nth) {
+    on_sides_[nth].reset(new Side_of_triangle_mesh(aabb_tree(nth)));
+  }
+
+  Side_of_triangle_mesh& on_side(int nth) {
+    if (!has_on_side(nth)) {
+      update_on_side(nth);
+    }
+    return *on_sides_[nth];
+  }
+
+  bool has_gps(int nth) { return gps_[nth] != nullptr; }
+
+  General_polygon_set_2& gps(int nth) {
+    if (!has_gps(nth)) {
+      gps_[nth].reset(new General_polygon_set_2);
+    }
+    return *gps_[nth];
+  }
+
+  bool has_pwh(int nth) { return pwh_[nth] != nullptr; }
+  Polygons_with_holes_2& pwh(int nth) {
+    if (!has_pwh(nth)) {
+      pwh_[nth].reset(new Polygons_with_holes_2);
+    }
+    return *pwh_[nth];
+  }
+
+  bool has_input_segments(int nth) { return input_segments_[nth] != nullptr; }
+
+  std::vector<Segment>& input_segments(int nth) {
+    if (!has_input_segments(nth)) {
+      input_segments_[nth].reset(new Segments);
+    }
+    return *input_segments_[nth];
+  }
+
+  bool has_segments(int nth) { return segments_[nth] != nullptr; }
+
+  Segments& segments(int nth) {
+    if (!has_segments(nth)) {
+      segments_[nth].reset(new Segments);
+    }
+    return *segments_[nth];
+  }
+
+  bool has_edges(int nth) { return edges_[nth] != nullptr; }
+
+  Edges& edges(int nth) {
+    if (!has_edges(nth)) {
+      edges_[nth].reset(new Edges);
+    }
+    return *edges_[nth];
+  }
+
+  bool has_input_points(int nth) { return input_points_[nth] != nullptr; }
+
+  Points& input_points(int nth) {
+    if (!has_input_points(nth)) {
+      input_points_[nth].reset(new Points);
+    }
+    return *input_points_[nth];
+  }
+
+  bool has_points(int nth) { return points_[nth] != nullptr; }
+
+  std::vector<Point>& points(int nth) {
```

Comment:
Is it normal to have this stuff in the the .h file? Not complaining, just trying to learn. I think of .h files as being definitions only

Reply:
Traditionally .h files just contain the interfaces, but with header libraries the definitions need to be there too.